text/0000-Element-Modifier-Managers.md

```diff
@@ -0,0 +1,444 @@
+    instance.element = element;
+    if (instance.wasInstalled !== undefined) {
+      instance.wasInstalled(args.positional, args.named);
+    }
+  },
+
+  updateModifier(instance, args) {
+    if (instance.didUpdateArguments !== undefined) {
+      instance.didUpdateArguments(args.positional, args.named);
+    }
+  }
+
+  // ...
+});
+```
+
+`updateModifier` recieves the modifier instance and also the the updated snapshot of arguments.
+
+This hook has the following timing semantics:
+
+**Always**
+
+- called after the arguments to the modifier have changed
+
+**Never**
+
+- called if the arguments to the modifier are constants
+
+### `destroyModifier`
+
+`destroyModifier` will be called when the modifier is no longer needed. This is intended for performing object-model level cleanup.
+
+```js
+// ember-basic-component/app/modifier-managers/basic.js
+
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  createModifier(factory, args) {
+    return factory.create(args.named);
+  },
+
+  installModifier(instance, element, args) {
+    instance.element = element;
+    if (instance.wasInstalled !== undefined) {
+      instance.wasInstalled(args.positional, args.named);
+    }
+  },
+
+  destroyModifier(instance, args) {
+    if (instance.willDestroyDOM !== undefined) {
+      instance.willDestroyDOM();
+    }
+  }
+});
+```
+
+This hook has the following timing semantics:
+
+**Always**
+
+- called after all children modifier manager's `destroyModifier` hook is called
+
+**May or May Not**
+
+- be called in the same tick as DOM removal
+
+## Capabilities
+
+In addition to the methods specified above, modifier managers are required to
+have a `capabilities` property.  This property must be set to the result of
+calling the `capabilities` function provided by Ember.
+
+### Versioning
+
+The first, mandatory, argument to the `capabilities` function is the modifier
+manager API, which is denoted in the `${major}.${minor}` format, matching the
+minimum Ember version this manager is targeting. For example:
+
+```js
+// ember-basic-component/app/component-managers/basic.js
+
+import { capabilities } from '@ember/modifier';
+import EmberObject from '@ember/object';
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.6'),
+
+  createModifier(factory, args) {
+    return factory.create(args.named);
+  },
+
+  installModifier(instance, element, args) {
+    instance.element = element;
+    if (instance.wasInstalled !== undefined) {
+      instance.wasInstalled(args.positional, args.named);
+    }
+  },
+
+  destroyModifier(instance, args) {
+    if (instance.willDestroyDOM !== undefined) {
+      instance.willDestroyDOM();
+    }
+  }
+});
+```
+
+This allows Ember to introduce new capabilities and make improvements to this
+API without breaking existing code.
+
+Here is a hypothical scenario for such a change:
+
+1. Ember 3.2 implemented and shipped the modifier manager API as described in
```

Comment:
Should this be 3.6 instead of 3.2? Seems to fit better with the rest of the hypotheticals below ...